main_test.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"testing"
+	"time"
+)
+
+var kernelImagePath = "./vmlinux"
+var rootDrivePath = "./root-drive.img"
+
+func init() {
+	if v := os.Getenv("KERNELIMAGE"); len(v) != 0 {
+		kernelImagePath = v
+	}
+
+	if v := os.Getenv("ROOTDRIVE"); len(v) != 0 {
+		rootDrivePath = v
+	}
+}
+
+func TestFireCTL(t *testing.T) {
+	const firectlName = "./firectl"
+	if _, err := os.Stat(firectlName); os.IsNotExist(err) {
+		t.Skip("Missing firectl binary")
+	}
+
+	if _, err := os.Stat(kernelImagePath); os.IsNotExist(err) {
+		t.Skip("Missing kernel image")
+	}
+
+	if _, err := os.Stat(rootDrivePath); os.IsNotExist(err) {
+		t.Skip("Missing root drive")
+	}
+
+	ctx := context.Background()
+	ctlCtx, cancelFn := context.WithTimeout(ctx, 500*time.Millisecond)
+	defer cancelFn()
+
+	const socketpath = "./integration_test.sock"
+	firectlArgs := []string{
+		"-s",
+		socketpath,
+		"--kernel",
+		kernelImagePath,
+		"--root-drive",
+		rootDrivePath,
+	}
+	cmd := exec.CommandContext(ctlCtx, firectlName, firectlArgs...)
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("Failed to run command: %v", err)
+	}
+
+	defer func() {
+		if err := os.Remove(socketpath); err != nil {
+			t.Log("Failed ot remove socket path file")
+		}
+
+		if err := cmd.Process.Kill(); err != nil {
+			t.Log("Failed to kill firectl process")
+		}
+	}()
+
+	// Sleep enough for firecracker to boot up
+	time.Sleep(500 * time.Millisecond)
+	tr := &http.Transport{
+		Dial: func(proto, addr string) (net.Conn, error) {
+			return net.Dial("unix", socketpath)
+		},
+	}
+	client := &http.Client{Transport: tr}
+	_, err := client.Get("http://foo")
```

Comment:
It may be overkill, but can you check for a proper InstanceInfo response? You may need to tweak the kernel's `panic` commandline setting to keep the VM from self-destructing long enough for you to do this.

Reply:
Yea, I was debating whether or not if I wanted to check the response. I'll go ahead and check to ensure that the status is running.

Reply:
You can also check that the returned state is something other than "Uninitialized", which indicates that we've successfully started a VM (that is, talked to the API) rather than just started the firecracker process.

Reply:
I am a little more hesitant to check for `Uninitialized` because we have to find the sweet spot of when the VM has been started but not the instance. I think it makes sense to wait for the instance to boot up, then hit it and check for running. What do you think?

Reply:
The instance doesn't need to finish booting; we just need to have send the InstanceStart  API, at which point firecracker should return a non-"Uninitialized" value. If you really wanted to, you could retry several times if you keep getting "Uninitialized", and timeout after some reasonably long delay.